internal/test/test.go

```diff
@@ -1,41 +1,157 @@
-func MockJSONRPCServer(requestChan chan *RequestData) (*httptest.Server, func(string)) {
-	var mu sync.RWMutex
-	// needed to retrieve requests that arrived at httpServer for further investigation
-	presetResponse := ""
-	setNextResponse := func(s string) {
-		mu.Lock()
-		defer mu.Unlock()
-		presetResponse = s
-	}
-
-	httpServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		data, _ := ioutil.ReadAll(r.Body)
-		defer r.Body.Close()
-		if requestChan != nil {
-			requestChan <- &RequestData{r, string(data)} // store the request for inspection
-		}
-		mu.RLock()
-		defer mu.RUnlock()
-		fmt.Fprintf(w, presetResponse) // respond with the preset response
-	}))
+// channel to a buffer size of 1. then writes to the chan will block until you read it. see
+// ReqChan() for how to do this
+func MockHTTPServer(requestChan chan *Request) *MockServer {
+	next := make(chan string, 1)
+	return &MockServer{
+		NextResponse: next,
+		Server: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			defer r.Body.Close()
+			if requestChan != nil {
+				data, _ := ioutil.ReadAll(r.Body)
+				requestChan <- &Request{r, w, string(data)}
+			}
+			fmt.Fprintf(w, <-next)
+		})),
+	}
+}
+
+// ReqChan makes a channel for reading received requests one by one.
+// Use it in conjunction with MockHTTPServer
+func ReqChan() chan *Request {
+	return make(chan *Request, 1)
+}
+
+// ReqToStr is a convenience method
+func ReqToStr(t *testing.T, req jsonrpc.RPCRequest) string {
+	r, err := json.Marshal(req)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return string(r)
+}
+
+// StrToReq is a convenience method
+func StrToReq(t *testing.T, req string) jsonrpc.RPCRequest {
+	var r jsonrpc.RPCRequest
+	err := json.Unmarshal([]byte(req), &r)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return r
+}
+
+// ResToStr is a convenience method
+func ResToStr(t *testing.T, res jsonrpc.RPCResponse) string {
+	r, err := json.Marshal(res)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return string(r)
+}
+
+func RandServerAddress(t *testing.T) string {
+	for _, addr := range config.GetLbrynetServers() {
+		return addr
+	}
+	t.Fatal("no lbrynet servers configured")
+	return ""
+}
+
+// JsonCompact removes insignificant space characters from a JSON string
+// It helps compare JSON strings without worrying about whitespace differences
+func JsonCompact(jsonStr string) string {
+	dst := &bytes.Buffer{}
+	err := json.Compact(dst, []byte(jsonStr))
+	if err != nil {
+		panic(err)
+	}
+	return dst.String()
+}
 
-	return httpServer, setNextResponse
+// JsonCompare compares two json strings.
+func jsonCompare(a, b []byte) (bool, string) {
+	opts := jsondiff.DefaultConsoleOptions()
+	diff, str := jsondiff.Compare(a, b, &opts)
+	return diff == jsondiff.FullMatch, str
+}
+
+// assert.Equal for JSON - more accurate comparison, pretty diff
+func AssertJsonEqual(t *testing.T, expected, actual interface{}, msgAndArgs ...interface{}) bool {
```

Comment:
This is going to save a lot of headache. Perhaps it would be a good idea to move out as a common package to lbry.go as it seems to suffer from "real world" test coverage because tests are not easy to write/run.

Reply:
not sure what you mean by real world test coverage. its using a library to do all the heavy lifting, then formatting it a bit nicer to match the assertion library we're using for the rest of the tests